### DISCUSSION TITLE
Support for terminal plotting

### DISCUSSION THREAD

**Discussion by Zethson** [Textualize/rich#1002] [Ideas]
Hey,

what I am really missing is a package which plots awesomely looking charts/plots in the terminal.
I would love to have a package with which I could not only display tables (thanks Rich!), but also bar charts, histograms, lineplots etc.

Inspiration: https://github.com/nschloe/termplotlib

Maybe this could be a nice extension of Rich in the future?

Eager to hear what you think!
Cheers

**Comment by willmcgugan** [comment]
That would be a nice to have feature, but it would be a lot to maintain, and frankly I’m not sure I know enough about such things to do it justice.

Ideally the author of termplot could write an integration for a Rich, so we can have the best of both worlds...

**Reply by zokrezyl** [reply]
https://yetty.dev  native shader (gpu) based plotting in the terminal. (zokrezyl/yetty on github)

**Comment by FedeClaudi** [comment]
Hi, 

this works:
```
from rich.layout import Layout
from rich import print
from rich.panel import Panel

import termplotlib as tpl
import numpy

x = numpy.linspace(0, 2 * numpy.pi, 10)
y = numpy.sin(x)

fig = tpl.figure()
fig.plot(x, y, label="data", width=50, height=15)

print(Panel(fig.get_string(), title='myplot'))
```

<img width="1420" alt="image" src="https://user-images.githubusercontent.com/17436313/107927645-2ce21880-6f6f-11eb-847a-e4fd6196f740.png">


Should be easy enough to edit `termplotlib` to have a `__rich_console__` method, will send a PR soon to @nschloe. Then we can have both things working together :)

**Reply by Zethson** [reply]
Maybe it would also be an even better idea to add this to https://github.com/mkaz/termgraph ?

**Reply by willmcgugan** [reply]
Integrations for both libraries would be amazing to have.

**Reply by FedeClaudi** [reply]
termgraph looks cool! I will have a play with it but it should be doable.

**Comment by nschloe** [comment]
Author of termplotlib here. tpl doesn't do any plotting itself but instead relies on [gnuplot](http://www.gnuplot.info/)'s terminal output. That's because it's _hard_ to plot on the terminal, and I concur with @willmcgugan that it would probably too much work to try and rewrite it in rich.

**Reply by FedeClaudi** [reply]
That's true, but termplot works out of the box with `rich`:
<img width="1792" alt="image" src="https://user-images.githubusercontent.com/17436313/107935655-9535f780-6f79-11eb-907d-051c11da1f0c.png">

e.g. with
```python
print(Panel(fig.get_string(), title='myplot'))
```

The only minor change would be to have `termplot.Figure` have a `__rich_console__` method which returns the string version of the plot, to avoid having to call `fig.get_string()`.

**Reply by willmcgugan** [reply]
Could you also specify the width of these plots? For instance, it might be nice if the "reward history" plot stretched to the available space.

**Reply by FedeClaudi** [reply]
You can, `termplot.Figure` has a height and width parameters. I left the space there to add another small panel ;)

**Reply by nschloe** [reply]
PR to tpl more than welcome!

**Comment by The-Compiler** [comment]
Another possible library to integrate (or just a source for inspiration) is [asciichartpy](https://pypi.org/project/asciichartpy/) (nevermind the JS-heavy readme, it's a Python port of a library which was originally written in JS). Its graphs are simple but pretty nice:

![image](https://user-images.githubusercontent.com/625793/109777710-4e483300-7c04-11eb-8944-08ae6fb74fec.png)

And it's code is [much simpler](https://github.com/kroitor/asciichart/blob/master/asciichartpy/__init__.py) than I expected (only some ~150 lines if you exclude a rather long docstring).

**Reply by willmcgugan** [reply]
That's rather nice. I'd love to see a port / integration with Rich. Up for a challenge @The-Compiler ? :-)

**Reply by The-Compiler** [reply]
Not at the moment I'm afraid - I personally wouldn't use it at the moment (I just got it from a friend and loved how it looks), and I'm afraid I have [more than enough on my plate](https://www.qutebrowser.org) myself :slightly_smiling_face:

**Reply by willmcgugan** [reply]
Fair enough. A keyboard powered browser? Very cool. I guess that makes it very quick to use once you have committed the keys to muscle memory?

**Reply by The-Compiler** [reply]
Yep - and once you get used to it, it's rather annoying to go back to using a browser "normally", similarly to how you end up with `i` scattered across your text files when you're used to vim but use a "normal" editor. That's why I originally started the project in 2013, because its [spiritual predecessor](https://wiki.archlinux.org/index.php/DWB) wasn't maintained anymore and there weren't any suitable alternatives around. :slightly_smiling_face: 

FWIW I passed the idea about integrating something like asciichartpy into rich on to said friend. :slightly_smiling_face:

**Reply by shusso** [reply]
It seems that `asciichartpy` works pretty nicely out-of-the-box with rich
```
import time
from rich.panel import Panel
from rich.live import Live

import asciichartpy as acp
import numpy as np


def simple_sine(s):
    return np.sin(2 * np.pi * (0.1 * s))

def get_panel(data):
    return Panel(acp.plot(data), expand=False, title="~~ [bold][yellow]waves[/bold][/yellow] ~~")

with Live(refresh_per_second=4) as live:
    d = []
    for i in range(1000):
        time.sleep(0.1)
        d.append(simple_sine(i))
        # pass only X latest 
        live.update(get_panel(d[-50:]))
```

**Comment by Abdur-rahmaanJ** [comment]
Released [termcharts](https://github.com/Abdur-rahmaanJ/termcharts) [[pypi](https://pypi.org/project/termcharts/)] with rich compatibility in mind.

**Reply by snth** [reply]
Your barcharts look very nice. They have that nice Rich aesthetic! Kudos!

**Reply by Abdur-rahmaanJ** [reply]
Thanks a lot!

**Reply by amal-khailtash** [reply]
Is there a way to add these charts to table cell/rows/columns

**Reply by Abdur-rahmaanJ** [reply]
@amal-khailtash The library above has paramter `rich=True` then you can use it as any rich element!

**Reply by amal-khailtash** [reply]
The question is how to add a chart to a rich table?  Do you have an example?

**Reply by Abdur-rahmaanJ** [reply]
@amal-khailtash Just like you add text, as the charts boil down to a bunch of text

**Comment by snth** [comment]
I just came here to see if termgraph or something similar was supported in Rich. Glad there are some options to try.

Here's another one to add to the list:
https://github.com/mkaz/termgraph

**Reply by NikosAlexandris** [reply]
Pity, it seems to not be maintained anymore.

**Comment by WildRackoon** [comment]
For anyone finding this thread:
`plotext` is probably the most up to date Terminal plotting lib as of today and has official `rich` support documentation:

https://github.com/piccolomo/plotext/blob/master/readme/environments.md#rich

**Comment by NikosAlexandris** [comment]
I think https://github.com/olavolav/uniplot is an amazing terminal plotting library / tool.
